src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2925,6 +2969,135 @@ void ProcessSystemTypeGetMemberLikeCall (System.Reflection.MemberTypes memberTyp
+						continue;
+
+					if ((bindingFlags & (BindingFlags.Instance | BindingFlags.Static)) == BindingFlags.Instance && method.IsStatic)
+						continue;
+
+					if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == BindingFlags.Public && !method.IsPublic)
+						continue;
+
+					if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == BindingFlags.NonPublic && method.IsPublic)
+						continue;
+
+					if (parametersCount != null && parametersCount != method.Parameters.Count)
+						continue;
+
+					ReportRecognizedReflectionAccessPattern (methodCalling, methodCalledDefinition, method);
+					Tracer.Push ($"Reflection-{method}");
+					try {
+						MarkIndirectlyCalledMethod (method);
+					} finally {
+						Tracer.Pop ();
+					}
+				}
+			}
+
+			void MarkPropertiesFromReflectionCall (TypeDefinition declaringType, string name, bool staticOnly = false)
+			{
+				foreach (var property in declaringType.Properties) {
+					if (property.Name != name)
+						continue;
+
+					Tracer.Push ($"Reflection-{property}");
+					try {
+						bool markedAny = false;
+
+						// It is not easy to reliably detect in the IL code whether the getter or setter (or both) are used.
+						// Be conservative and mark everything for the property.
+						var getter = property.GetMethod;
+						if (getter != null && (!staticOnly || staticOnly && getter.IsStatic)) {
+							MarkIndirectlyCalledMethod (getter);
+							markedAny = true;
+						}
+
+						var setter = property.SetMethod;
+						if (setter != null && (!staticOnly || staticOnly && setter.IsStatic)) {
+							MarkIndirectlyCalledMethod (setter);
+							markedAny = true;
+						}
+
+						if (markedAny) {
+							ReportRecognizedReflectionAccessPattern (methodCalling, methodCalledDefinition, property);
+							MarkProperty (property);
+						}
+
+					} finally {
+						Tracer.Pop ();
+					}
+				}
+			}
+
+			void MarkFieldsFromReflectionCall (TypeDefinition declaringType, string name, bool staticOnly = false)
+			{
+				foreach (var field in declaringType.Fields) {
+					if (field.Name != name)
+						continue;
+
+					if (staticOnly && !field.IsStatic)
+						continue;
+
+					ReportRecognizedReflectionAccessPattern (methodCalling, methodCalledDefinition, field);
+
+					Tracer.Push ($"Reflection-{field}");
+					try {
+						MarkField (field);
+					} finally {
+						Tracer.Pop ();
+					}
+
+					break;
+				}
+			}
+
+			void MarkEventsFromReflectionCall (TypeDefinition declaringType, string name)
+			{
+				foreach (var eventInfo in declaringType.Events) {
+					if (eventInfo.Name != name)
+						continue;
+
+					ReportRecognizedReflectionAccessPattern (methodCalling, methodCalledDefinition, eventInfo);
+
+					Tracer.Push ($"Reflection-{eventInfo}");
+					try {
+						MarkEvent (eventInfo);
+					} finally {
+						Tracer.Pop ();
+					}
+				}
+			}
+
+			void ReportUnrecognizedReflectionCallPattern (string message)
+			{
+				reportedPatternRecognition = true;
```

Comment:
I don't like this approach. It might be a neat way to leverage compiler-generated wrapper class but I'd much prefer to move whole reflection detection logic to worked like struct and make it explicit there

Reply:
I adapted the approach suggested by @mrvoorhe. Could not make it a struct all the way as the suggested pattern of actions requires a ref-type so that it can be passed to the closure.
But this way it should only allocate one class per method body and the per-call info is a struct passed by reference, so doesn't allocate on heap.